From fc35dd95930a0e6c70e0d9d891c6bab36f096ed7 Mon Sep 17 00:00:00 2001
From: Pushpal Sidhu <psidhu@gateworks.com>
Date: Fri, 10 Jul 2015 12:45:55 -0700
Subject: [PATCH] imxv4l2videosrc: v4l2_buffer_pool: Add special case for
 tw6869 driver

The tw6869 driver requires a different physical address to allow DMA'able
buffers for decreasing latency between chip and display.

Signed-off-by: Pushpal Sidhu <psidhu@gateworks.com>
---
 src/v4l2src/v4l2_buffer_pool.c | 14 +++++++++++---
 1 file changed, 11 insertions(+), 3 deletions(-)

diff --git a/src/v4l2src/v4l2_buffer_pool.c b/src/v4l2src/v4l2_buffer_pool.c
index 9146c32..b5fe8a9 100644
--- a/src/v4l2src/v4l2_buffer_pool.c
+++ b/src/v4l2src/v4l2_buffer_pool.c
@@ -137,6 +137,7 @@ static GstFlowReturn gst_imx_v4l2_buffer_pool_alloc_buffer(GstBufferPool *bpool,
 	GstImxV4l2Meta *meta;
 	GstImxPhysMemMeta *phys_mem_meta;
 	GstVideoInfo *info;
+	struct v4l2_capability cap = {0};
 
 	buf = gst_buffer_new();
 	if (buf == NULL)
@@ -166,10 +167,17 @@ static GstFlowReturn gst_imx_v4l2_buffer_pool_alloc_buffer(GstBufferPool *bpool,
 	g_assert(meta->mem);
 
 	phys_mem_meta = GST_IMX_PHYS_MEM_META_ADD(buf);
-	phys_mem_meta->phys_addr = meta->vbuffer.m.offset;
 
-	/* Safeguard to catch data loss if in any future i.MX version the types do not match */
-	g_assert(meta->vbuffer.m.offset == (__u32)(phys_mem_meta->phys_addr));
+	/* XXX: This is only required for the tw6869 driver */
+	ioctl(GST_IMX_FD_OBJECT_GET_FD(pool->fd_obj_v4l), VIDIOC_QUERYCAP, &cap);
+	if (strncmp(cap.card, "tw6869", 6) == 0)
+		phys_mem_meta->phys_addr = (*(__u32 *)meta->mem);
+	else {
+		phys_mem_meta->phys_addr = meta->vbuffer.m.offset;
+
+		/* Safeguard to catch data loss if in any future i.MX version the types do not match */
+		g_assert(meta->vbuffer.m.offset == (__u32)(phys_mem_meta->phys_addr));
+	}
 
 	if (pool->add_videometa)
 	{
-- 
2.4.3

